server/msgpack/server.go

```diff
@@ -96,17 +103,18 @@ func (s *handler) Handle(conn net.Conn) {
 		metric := it.Metric()
 		policiesList := it.PoliciesList()
 		if err := s.aggregator.AddMetricWithPoliciesList(metric, policiesList); err != nil {
-			// We sample the error log here because the error rate may scale with
+			// We rate limit the error log here because the error rate may scale with
 			// the metrics incoming rate and consume lots of cpu cycles.
-			// TODO(xichen): need to sample this in case there's a large number of errors
-			// due to rate limiting etc.
-			if s.rand.Float64() < s.errLogSamplingRate {
-				s.log.WithFields(
-					log.NewField("metric", metric.String()),
-					log.NewField("policies", policiesList),
-					log.NewErrField(err),
-				).Errorf("error adding metric with policies")
+			if s.errLogRateLimiter != nil && !s.errLogRateLimiter.IsAllowed(1) {
+				s.metrics.errLogRateLimited.Inc(1)
+				continue
 			}
+			s.log.WithFields(
+				log.NewField("type", metric.Type.String()),
+				log.NewField("id", metric.ID.String()),
+				log.NewField("policies", policiesList),
+				log.NewErrField(err),
+			).Errorf("error adding metric with policies")
 			s.metrics.addMetricErrors.Inc(1)
```

Comment:
nit: should we increment this even if we don't log the error? Otherwise we have to sum this and `errLogRateLimited ` which may be unintuitive.

Reply:
Right yeah makes sense.